File: 75-sort-colors/sort-colors.java
class Solution {
    public void sortColors(int[] nums) {
        int red = 0, white = 0, blue = nums.length - 1;
        
        while (white <= blue) {
            if (nums[white] == 0) {
                swap(nums, red++, white++);
            } else if (nums[white] == 1) {
                white++;
            } else {
                swap(nums, white, blue--);
            }
        }
    }
    
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 75-sort-colors/sort-colors.cpp
class Solution {
public:
    void sortColors(std::vector<int>& nums) {
        int low = 0;
        int high = nums.size() - 1;
        int mid = 0;
        
        while (mid <= high) {
            switch(nums[mid]) {
                case 0:
                    std::swap(nums[low++], nums[mid++]);
                    break;
                case 1:
                    mid++;
                    break;
                case 2:
                    std::swap(nums[mid], nums[high--]);
                    break;
            }
        }
    }
};